hoomd/write/hdf5.py

```diff
@@ -0,0 +1,295 @@
+                                 dtype, chunk_size)
+
+    @_skip_fh
+    def _find_frame(self):
+        if "hoomd-data" in self._fh:
+            return self._fh["hoomd-data"].attrs["frames"]
+        return 0
+
+    @_skip_fh
+    def _validate_scheme(self):
+        if "hoomd-data" in self._fh:
+            if "hoomd-schema" not in self._fh["hoomd-data"].attrs:
+                raise RuntimeError("Validation of existing HDF5 file failed.")
+        else:
+            group = self._fh.create_group("hoomd-data")
+            group.attrs["hoomd-schema"] = [0, 1]
+            group.attrs["frames"] = 0
+
+    def __getstate__(self):
+        state = copy.copy(self.__dict__)
+        del state["_fh"]
+        state["_attached_"] = False
+        return state
+
+    def __setstate__(self, state):
+        self.__dict__ = state
+
+
+class HDF5Logger(_InternalCustomWriter):
+    """An HDF5 logger backend.
+
+    This class handles scalar and array data storing them in HDF5 resizable
+    datasets.
+
+    Note:
+        This class requires that ``h5py`` be installed.
+
+    Important:
+        The HDF5 file can be used for other data storage; however, the
+        "hoomd-data" key is reserved for use by this class. An exception will be
+        thrown if this requirement is not met.
+
+    Warning:
+        This class cannot handle string, strings, or object loggables.
+
+    Args:
+        trigger (hoomd.trigger.trigger_like): The trigger to determine when to
+            run the HDF5 backend.
+        filename (str): The filename of the HDF5 file to write to.
+        logger (hoomd.logging.Logger): The logger instance to use for querying
+            log data.
+        mode (`str`, optional): The mode to open the file in. Available values
+            are "w", "x" and "w-", "a", and "r+". Defaults to "a". See the
+            h5py_ documentation for more details).
+
+    .. _h5py:
+        https://docs.h5py.org/en/stable/high/file.html#opening-creating-files
+
+    Attributes:
+        cls.accepted_categories (hoomd.logging.LoggerCategories): The enum value
+            for all accepted categories for `HDF5Logger` instances which is all
+            categories other than "string", "strings", and "object" (see
+            `hoomd.logging.LoggerCategories`).
+        trigger (hoomd.trigger.trigger_like): The trigger to determine when to
+            run the HDF5 backend.
+        filename (str): The filename of the HDF5 file written to.
+        logger (hoomd.logging.Logger): The logger instance used for querying
+            log data.
+        mode (str): The mode the file was opened in.
+
+    .. rubric:: Example
+
+    .. invisible-code-block: python
+
+        import hoomd
+
+        try:
+            import h5py
+            h5py_not_available = False
+        except ImportError:
+            h5py_not_available = True
+
+
+        simulation = hoomd.util.make_example_simulation()
+        h5_filename = tmp_path / "simulation_log.h5"
```

Comment:
```suggestion
        hdf5_filename = tmp_path / "simulation_log.h5"
```
Use the same name throughout. We don't abbreviate with h5 anywhere else. Alternately, use H5 everywhere: e.g. name the class `H5Log` to be consistent with the file extension.